docs/proposals/kubegen.md

```diff
@@ -0,0 +1,214 @@
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - start license with "Copyright <current year> <copyright>" e.g. *Copyright 2018 The Kubernetes Authors.*
+  - exit non-zero if the LICENSE file is missing
+
+
+#### Specifying the license file
+
+The license file can be overridden from the LICENSE file using the `--license-file` flag
+
+```sh
+kubegen --license-file "boilerplate.go.txt"
+```
+
+- run all code generators against discovered APIs
+  - run for both internal and external types
+- prepend license to generated files
+  - use the contents from boilerplate.go.txt
+  - if boilerplate.go.txt is already in comments then use it verbatim, otherwise wrap it in comments
+  - exit non-zero if boilerplate.go.txt is missing
+
+#### Specifying a license type
+
+The license can be overridden from the LICENSE file using the `--license` flag
+
+```sh
+kubegen --license "Apache 2.0"
+```
+
+- run all code generators against discovered APIs
+  - run for both internal and external types
+- prepend license to generated files
+  - use an *Apache 2.0* license
+  
+#### Generating for external types only
+
+Generating code for internal types can be skipped using the `generate-internal` flag
+
+```sh
+kubegen --generate-internal=false
+```
+
+- run all code generators against discovered APIs
+  - run for external types only
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - exit non-zero if the LICENSE file is missing
+
+#### Specifying API versions
+
+Generating code only for specific API group versions can be performed by providing the API
+group versions as positional arguments.
+
+```sh
+kubegen apps/v1 apps/v1beta1 extensions/v1beta1
+```
+
+- run all code generators against the apps/v1 apps/v1beta1 extensions/v1beta1 API groups versions
+  - run for both internal and external types
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - exit non-zero if the LICENSE file is missing
+
+#### Specifying generators
+
+Generating code using only a specific set of generators can be performed using the
+`--generator` flag.  This flag may be provided multiple times to run multiple
+generators.
+
+```sh
+kubegen --generator client-gen --generator lister-gen
+```
+
+- run only the client-gen and lister-gen code generators
+  - run for both internal and external types
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - exit non-zero if the LICENSE file is missing
+  
+#### Specifying where to search for APIs
+
+Looking for APIs outside the default location can be configured using
+the `--apis-dir` flag.  This flag may be provided multiple times to search multiple
+directories.
+
+```sh
+kubegen --apis-dir notpkg/apis --apis-dir pkg/notapis
+```
+
+- run all code generators against discovered APIs
+  - search for API group versions under notpkg/apis and pkg/notapis instead of pkg/apis  
+
```

Comment:
What about the output-package?

Reply:
Do we pass through any gengo args?

Reply:
For reference, these are the gengo flags:
```
	fs.StringSliceVarP(&g.InputDirs, "input-dirs", "i", g.InputDirs, "Comma-separated list of import paths to get input types from.")
	fs.StringVarP(&g.OutputBase, "output-base", "o", g.OutputBase, "Output base; defaults to $GOPATH/src/ or ./ if $GOPATH is not set.")
	fs.StringVarP(&g.OutputPackagePath, "output-package", "p", g.OutputPackagePath, "Base package path.")
	fs.StringVarP(&g.OutputFileBaseName, "output-file-base", "O", g.OutputFileBaseName, "Base name (without .go suffix) for output files.")
	fs.StringVarP(&g.GoHeaderFilePath, "go-header-file", "h", g.GoHeaderFilePath, "File containing boilerplate header text. The string YEAR will be replaced with the current 4-digit year.")
	fs.BoolVar(&g.VerifyOnly, "verify-only", g.VerifyOnly, "If true, only verify existing output, do not write anything.")
	fs.StringVar(&g.GeneratedBuildTag, "build-tag", g.GeneratedBuildTag, "A Go build tag to use to identify files generated by this command. Should be unique.")
```

Reply:
There might be generator-specific flags we have to pass through as well. We should have a way to do this and/or look through the existing gen-specific flags.

Reply:
Agreed, though I would prefer to initially provide the simplest interface that works, and then add this additional complexity as we discover it is needed.

Reply:
Lets start with defaulting the output package and add it when we discover it is necessary.

Reply:
Actually, it's only `--verify-only` I care about right now from the generic flags. But there are more converter specific flags. Just saying that we should be prepared to offer converter specific ones, e.g. `--conversion-gen-exceptions-file`, `--client-gen-clientset-name`.

Reply:
SGTM.  I am proposing exposing `--verify-only` as `--dry-run` since it is a much more common flag name and has the same meaning.  Thoughts?